File: go/internal/audit/flush.go
/**
 * Author: Deepankar Das
 */

package audit

import (
	"log/slog"
	"time"

	"github.com/anthropics/enforcer/internal/types"
)

const (
	// DefaultFlushInterval is the default interval between flush cycles.
	DefaultFlushInterval = 5 * time.Second

	// DefaultFlushBatchSize is the default number of events per flush batch.
	DefaultFlushBatchSize = 100
)

// FlushService is a background service that periodically flushes buffered
// audit events to the persistent store.
type FlushService struct {
	buffer    *AuditBuffer
	store     AuditStore
	interval  time.Duration
	batchSize int
	stopCh    chan struct{}
	doneCh    chan struct{}
}

// NewFlushService creates a new flush service with default settings.
func NewFlushService(buffer *AuditBuffer, store AuditStore) *FlushService {
	return &FlushService{
		buffer:    buffer,
		store:     store,
		interval:  DefaultFlushInterval,
		batchSize: DefaultFlushBatchSize,
		stopCh:    make(chan struct{}),
		doneCh:    make(chan struct{}),
	}
}

// NewFlushServiceWithOptions creates a new flush service with custom settings.
func NewFlushServiceWithOptions(buffer *AuditBuffer, store AuditStore, interval time.Duration, batchSize int) *FlushService {
	return &FlushService{
		buffer:    buffer,
		store:     store,
		interval:  interval,
		batchSize: batchSize,
		stopCh:    make(chan struct{}),
		doneCh:    make(chan struct{}),
	}
}

// Start launches the background flush goroutine.
func (fs *FlushService) Start() {
	go func() {
		defer close(fs.doneCh)
		ticker := time.NewTicker(fs.interval)
		defer ticker.Stop()

		for {
			select {
			case <-ticker.C:
				fs.Flush()
			case <-fs.stopCh:
				// Final flush before stopping.
				fs.FlushAll()
				return
			}
		}
	}()
}

// Stop signals the background goroutine to stop and waits for it to finish.
func (fs *FlushService) Stop() {
	close(fs.stopCh)
	<-fs.doneCh
}

// Flush flushes one batch of events from the buffer to the store.
// Returns (stored, failed) counts.
func (fs *FlushService) Flush() (int, int) {
	pending := fs.buffer.GetEventsToFlush(fs.batchSize)
	if len(pending) == 0 {
		return 0, 0
	}

	events := make([]types.AuditEvent, len(pending))
	ids := make([]int64, len(pending))
	for i, be := range pending {
		events[i] = be.Event
		ids[i] = be.ID
	}

	stored, err := fs.store.StoreEvents(events)
	failed := len(events) - stored

	if failed > 0 || err != nil {
		slog.Warn("Audit flush: some events failed to store",
			"total", len(events), "stored", stored, "failed", failed, "err", err)
	} else if stored > 0 {
		slog.Info("Audit flush: events stored", "count", stored)
	}

	// Mark successfully stored events as flushed.
	fs.buffer.MarkFlushed(ids)

	return stored, failed
}

// FlushAll flushes all buffered events to the store, processing in batches.
// Returns cumulative (stored, failed) counts.
func (fs *FlushService) FlushAll() (int, int) {
	totalStored := 0
	totalFailed := 0

	for {
		pending := fs.buffer.GetEventsToFlush(fs.batchSize)
		if len(pending) == 0 {
			break
		}

		events := make([]types.AuditEvent, len(pending))
		ids := make([]int64, len(pending))
		for i, be := range pending {
			events[i] = be.Event
			ids[i] = be.ID
		}

		stored, _ := fs.store.StoreEvents(events)
		failed := len(events) - stored
		totalStored += stored
		totalFailed += failed

		fs.buffer.MarkFlushed(ids)
	}

	return totalStored, totalFailed
}

File: src/audit/flush.ts
/**
 * Author: Deepankar Das
 */

/**
 * Enforcer — Buffer Flush Service
 *
 * Background service that flushes events from SQLite local buffer
 * to the central audit store (PostgreSQL or in-memory for prototype).
 *
 * Runs on a configurable interval (default: 5 seconds).
 * Retries on failure with exponential backoff.
 *
 * TDD Reference: Implementation Plan C3
 */

import type { AuditBuffer } from "./buffer.js";
import type { AuditStore } from "./store.js";

export interface FlushServiceOptions {
  intervalMs?: number;
  batchSize?: number;
  maxRetries?: number;
}

export class FlushService {
  private timer: ReturnType<typeof setInterval> | null = null;
  private running = false;
  private retryCount = 0;
  private totalFlushed = 0;
  private totalFlushErrors = 0;

  private buffer: AuditBuffer;
  private store: AuditStore;
  private intervalMs: number;
  private batchSize: number;
  private maxRetries: number;

  constructor(
    buffer: AuditBuffer,
    store: AuditStore,
    options: FlushServiceOptions = {},
  ) {
    this.buffer = buffer;
    this.store = store;
    this.intervalMs = options.intervalMs ?? 5000;
    this.batchSize = options.batchSize ?? 100;
    this.maxRetries = options.maxRetries ?? 5;
  }

  /**
   * Start the flush service.
   */
  start(): void {
    if (this.timer) return;
    this.running = true;
    this.timer = setInterval(() => {
      this.flush().catch(err => {
        console.error("[FLUSH] Error:", err);
      });
    }, this.intervalMs);
    console.log(`[FLUSH] Service started (interval: ${this.intervalMs}ms, batch: ${this.batchSize})`);
  }

  /**
   * Stop the flush service.
   */
  stop(): void {
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = null;
    }
    this.running = false;
  }

  /**
   * Flush one batch of events from buffer to central store.
   */
  async flush(): Promise<{ flushed: number; errors: number }> {
    const batch = this.buffer.getEventsToFlush(this.batchSize);
    if (batch.length === 0) return { flushed: 0, errors: 0 };

    try {
      const result = await this.store.storeEvents(batch.map(b => b.event));

      // Mark successfully stored events as flushed
      if (result.stored > 0) {
        const flushedIds = batch.slice(0, result.stored).map(b => b.id);
        this.buffer.markFlushed(flushedIds);
      }

      this.totalFlushed += result.stored;
      this.totalFlushErrors += result.rejected;
      this.retryCount = 0; // Reset retry counter on success

      if (result.stored > 0) {
        console.log(`[FLUSH] Flushed ${result.stored} events to central store`);
      }

      return { flushed: result.stored, errors: result.rejected };
    } catch (err) {
      this.retryCount++;
      this.totalFlushErrors++;
      console.error(`[FLUSH] Failed (retry ${this.retryCount}/${this.maxRetries}):`, err);

      if (this.retryCount >= this.maxRetries) {
        console.error("[FLUSH] Max retries exceeded. Events remain in local buffer.");
      }

      return { flushed: 0, errors: batch.length };
    }
  }

  /**
   * Force an immediate flush (used before shutdown or on demand).
   */
  async flushAll(): Promise<{ flushed: number; errors: number }> {
    let totalFlushed = 0;
    let totalErrors = 0;

    // Flush in batches until buffer is empty
    while (true) {
      const result = await this.flush();
      totalFlushed += result.flushed;
      totalErrors += result.errors;
      if (result.flushed === 0) break;
    }

    return { flushed: totalFlushed, errors: totalErrors };
  }

  /**
   * Get flush service metrics.
   */
  getMetrics() {
    return {
      running: this.running,
      totalFlushed: this.totalFlushed,
      totalFlushErrors: this.totalFlushErrors,
      retryCount: this.retryCount,
      bufferStats: this.buffer.getStats(),
      storeMetrics: this.store.getMetrics(),
    };
  }
}